### ISSUE TITLE
How to tune parameters & Estimating normal vectors and curvature before executing functions

### ISSUE THREAD

**Issue by TakuOkawara** [hanzheteng/color_icp#2]
Thank you for giving us wonderful tools. I have been applying this system to our research project.
I have two questions.

1. Could you give me some advice about how to tune parameters? 
My pcd's dataset is [here](https://drive.google.com/drive/folders/1DvO8nhx1NGEGn-UZTeYeXdp64urKwmvk?usp=share_link). And then, mismatch1.pcd and mismatch2.pcd are input to the function through [param.yaml file](https://drive.google.com/file/d/12vYF4WdEWZExfftHmLrzsHNhKr-VX2K_/view?usp=share_link).
Even if Optimization has been converged, some error (like doubled object) is included as seen in the following image.
![Screenshot from 2023-02-07 21-37-06](https://user-images.githubusercontent.com/105478884/217246800-39fad6ac-c12a-4c39-a2fe-1650e392fd72.png)
Therefore, I would like to ask you how to tune some parameters to correct this error.

2. Do I estimate a normal vector and curvature of the input point cloud before inputting the point cloud?
Some warnings occur when I execute color_icp. My dataset does not include a normal vector and curvature.
![Screenshot from 2023-02-07 21-44-33](https://user-images.githubusercontent.com/105478884/217248391-e374e9b8-d5d3-4910-9cca-333612c7437d.png)
However, these parameters seem to be calculated in estimateNormals() function located at 94 line, color_icp.cpp. Therefore, do I need to estimate a normal vector and curvature of the input point cloud before inputting the point cloud?

**Comment by hanzheteng** [comment]
I think you happen to show a case that `color_icp` works worse than the regular icp registration algorithm. I tried your data using the python script in this repo, and the result shows that the regular ICP works well, but the color_icp produces inaccurate registration results. See below the three results for comparison.

1. the original layout of the source and target point cloud.
![1](https://user-images.githubusercontent.com/22992305/217393005-440cc5b1-70ad-43bc-b37f-9d6c716c8973.png)

2. registration result by ICP algorithm
![2](https://user-images.githubusercontent.com/22992305/217393095-7b1aaba0-9731-424a-81e0-8f881d6c6936.png)

3. registration result by the color_icp algorithm
![3](https://user-images.githubusercontent.com/22992305/217393128-fdfd8e9e-4081-4f73-ba6b-9281aa351f67.png)

I guess the reason behind this is that the pair of cloud you provided contains more geometric clues than color ones. For example, the shape outline is clear, but the color texture is complicated, without a clear/unique "gradient" direction for the algorithm to optimize towards. I think the ideal target environment for color_icp algorithm to work is a flat surface with rich and clear color gradient.

**Comment by hanzheteng** [comment]
1. Some tuning suggestions:
- You may decrease the `voxel_resolution`, as in your case this is really a local scene and each centimeter matters, but the default voxel resolution is at 10cm level. I can tell this also from the number of points left after downsampling (original 120000, after 3500).
- If you change voxel size/resolution, you also need to change `normal_est_radius`, `search_radius`, and `icp_max_corres_dist` accordingly, as they are meant to be a few times larger than `voxel_resolution`, such that a reasonable number of points can be found in the local sphere neighborhood.
- You may try different `color_icp_lambda` value (1--> no color info integrated, 0-->all color info).
- You may try different `color_weight` method, though I don't expect it to provide non-trivial difference.


2. No, you don't need to worry about these warnings. The warning happens because the default point type in this algorithm is `pcl::PointXYZRGBNormal` (this point type is defined [here](https://github.com/PointCloudLibrary/pcl/blob/master/common/include/pcl/impl/point_types.hpp#L892), and we use them [here](https://github.com/hanzheteng/color_icp/blob/main/src/color_icp.cpp#L30)), and the function `pcl::io::loadPCDFile` (in [this line](https://github.com/hanzheteng/color_icp/blob/main/include/color_icp/helper.h#L167) and [this line](https://github.com/hanzheteng/color_icp/blob/main/src/color_icp.cpp#L84) of code) was trying to load from input clouds every possible field to match the designated point type. It couldn't find normal_x/y/z and curvature in the original input clouds, so it will leave them empty and produce this warning. This doesn't matter because these fields (normal_x/y/z and curvature) will be filled later on in the code as you mentioned, by the normal estimation algorithm and so on.

**Comment by TakuOkawara** [comment]
Thank you for your quick and polite comments.

I confirmed the regular ICP is more accurate than the color ICP thanks to your explanation and three results for comparison. Our actual environment is different from my data set attached in the previous post. This data set was just trial data. Furthermore, I understood how to tune parameters thanks to your polite explanation. I also follow the setting of normal vectors and curvature.

To sum up, your algorithm can be applied to our research based on your comments. Thank you!

**Comment by hanzheteng** [comment]
Thanks for your kind words. Glad you find it helpful and best of luck to your research :)

**Comment by gitouni** [comment]
Actually, this problem also happens in the [script](https://github.com/hanzheteng/color_icp/blob/main/scripts/colored_icp.py). ICP performs better than colored_ICP in registration between [color_bin_00_easy.pcd](https://github.com/hanzheteng/color_icp/blob/main/data/cloud_bin_00_easy.pcd) and [color_bin_01.pcd](https://github.com/hanzheteng/color_icp/blob/main/data/cloud_bin_01.pcd). Your comment makes me convinced that there are some scenarios not proper for colored method, thanks.

Note: you can use the following json setting to view the Open3D's view (Ctrl + V on the view window). Manifest difference can be found on the boundary of the armchair.
```json
{
	"class_name" : "ViewTrajectory",
	"interval" : 29,
	"is_loop" : false,
	"trajectory" : 
	[
		{
			"boundingbox_max" : [ 3.43359375, 2.7008078098297119, 2.1282944679260254 ],
			"boundingbox_min" : [ -0.30465653538703918, 0.89872455596923828, 0.47970101237297058 ],
			"field_of_view" : 60.0,
			"front" : [ 0.72311997525859728, -0.12254874450305511, -0.67976415513230082 ],
			"lookat" : [ 1.5644686073064804, 1.7997661828994751, 1.303997740149498 ],
			"up" : [ -0.27748035007253208, -0.95277190584179838, -0.12341130727049508 ],
			"zoom" : 0.099999999999999617
		}
	],
	"version_major" : 1,
	"version_minor" : 0
}
```
